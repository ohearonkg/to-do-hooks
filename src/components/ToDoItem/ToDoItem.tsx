import React from "react";

export interface IToDo {
  /**
   * The id for this to do
   */
  id: string;

  /**
   * Text to be displayed
   */
  text: string;
}

export interface IToDoItem extends IToDo {
  /**
   * Function to be called
   * when the user clicks upon
   * this item
   */
  onClickFunction: (toDoId: string) => void;
}

const ToDoItem: React.SFC<IToDoItem> = ({ text, onClickFunction, id }) => (
  <div onClick={() => onClickFunction(id)}>{text}</div>
);

export default ToDoItem;
