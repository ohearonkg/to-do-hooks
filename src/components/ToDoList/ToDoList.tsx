import ToDoItem, { IToDo } from "../ToDoItem/ToDoItem";

import React from "react";

interface IToDoList {
  /**
   * List of to do items
   */
  toDos: IToDo[];

  /**
   * Function to be called
   * when the user clicks
   * on a to do item
   */
  itemSelectedFunction: (selectedId: string) => void;
}
const ToDoList: React.SFC<IToDoList> = ({ toDos, itemSelectedFunction }) => (
  <div>
    {toDos.map((toDo: IToDo) => (
      <ToDoItem
        key={toDo.id}
        {...toDo}
        onClickFunction={itemSelectedFunction}
      />
    ))}
  </div>
);

export default ToDoList;
