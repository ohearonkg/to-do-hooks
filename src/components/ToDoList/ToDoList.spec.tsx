import "jest-dom/extend-expect";

import { cleanup, fireEvent, render } from "react-testing-library";

import React from "react";
import { IToDo } from "../ToDoItem/ToDoItem";
import ToDoList from "./ToDoList";

afterEach(cleanup);

/**
 * Variables for testing
 */
const sampleItemSelectedFunction = () => ({});

const sampleToDos: IToDo[] = [
  {
    id: "1",
    text: "a"
  },
  {
    id: "2",
    text: "b"
  }
];

describe("To Do List", () => {
  /**
   * Snapshot
   */
  it("Should match its snapshot", () => {
    const { asFragment } = render(
      <ToDoList
        toDos={sampleToDos}
        itemSelectedFunction={sampleItemSelectedFunction}
      />
    );
    expect(asFragment()).toThrowErrorMatchingSnapshot();
  });

  /**
   * Rendering <ToDoItem />
   */
  it("Should render a <ToDoItem /> for each item passed to its toDos prop", () => {
    const { getByText } = render(
      <ToDoList
        toDos={sampleToDos}
        itemSelectedFunction={sampleItemSelectedFunction}
      />
    );
    sampleToDos.forEach((toDo: IToDo) =>
      expect(getByText(toDo.text)).toBeInTheDocument()
    );
  });

  /**
   * Calling its itemSelectedFunction
   */
  it("Should call its itemSelectedFunction when the user clicks upon a <ToDoItem />", () => {
    const sampleFunction = jest.fn();
    const { getByText } = render(
      <ToDoList toDos={sampleToDos} itemSelectedFunction={sampleFunction} />
    );
    sampleToDos.forEach((toDo: IToDo, index: number) => {
      fireEvent.click(getByText(toDo.text));
      expect(sampleFunction).toHaveBeenCalledTimes(index + 1);
    });
  });

  it("Should call its itemSeletedFunction with the id of the todo clicked", () => {
    const sampleFunction = jest.fn();
    const { getByText } = render(
      <ToDoList toDos={sampleToDos} itemSelectedFunction={sampleFunction} />
    );
    sampleToDos.forEach((toDo: IToDo) => {
      fireEvent.click(getByText(toDo.text));
      expect(sampleFunction).toHaveBeenLastCalledWith(toDo.id);
    });
  });
});
