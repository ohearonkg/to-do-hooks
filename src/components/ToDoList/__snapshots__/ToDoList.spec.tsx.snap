// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`To Do List Should match its snapshot 1`] = `"received is not a function"`;
